src/aiai_eval/named_entity_recognition.py

```diff
@@ -85,6 +74,131 @@ def _preprocess_data_transformer(
             ]
         )
 
+    def _preprocess_data_spacy(self, dataset: Dataset) -> Dataset:
+        """Preprocess the given Huggingface dataset for use by a SpaCy model.
+
+        Args:
+            dataset (Dataset): The dataset to preprocess.
+
+        Returns:
+            Dataset:
+                The preprocessed dataset.
+        """
+        # Add a labels column to the dataset
+        def create_label_col(example):
+            example["labels"] = [
+                self.task_config.id2label[x] for x in example["labels"]
+            ]
+            return example
+
+        dataset = dataset.add_column("labels", dataset["ner_tags"])
+        return dataset.map(create_label_col)
+
+    def _get_spacy_predictions_and_labels(
+        self, model: Any, dataset: Dataset, batch_size: int
+    ) -> tuple:
+        """Get predictions from SpaCy model on dataset.
+
+        Args:
+            model (SpaCy model):
+                The model.
+            dataset (Hugging Face dataset):
+                The dataset.
+            batch_size (int):
+                The batch size to use.
+
+        Returns:
+            A pair of arrays:
+                The first array contains the probability predictions and the second
+                array contains the true labels.
+        """
+        # Initialise progress bar
+        if self.evaluation_config.progress_bar:
+            itr = tqdm(
+                dataset[self.task_config.feature_column_name],
+                desc="Evaluating model",
+                leave=False,
+            )
+        else:
+            itr = dataset[self.task_config.feature_column_name]
+
+        processed = model.pipe(itr, batch_size=batch_size)
+        map_fn = self._extract_spacy_predictions
+        predictions = map(map_fn, zip(dataset["tokens"], processed))
+
+        return list(predictions), dataset["labels"]
+
+    def _extract_spacy_predictions(self, tokens_processed: tuple) -> list:
+        """Helper function that extracts the predictions from a SpaCy model.
+        Aside from extracting the predictions from the model, it also aligns the
+        predictions with the gold tokens, in case the SpaCy tokeniser tokenises the
+        text different from those.
+        Args:
+            tokens_processed (tuple):
+                A pair of the labels, being a list of strings, and the SpaCy processed
+                document, being a Spacy `Doc` instance.
+        Returns:
+            list:
+                A list of predictions for each token, of the same length as the gold
+                tokens (first entry of `tokens_processed`).
+        """
```

Comment:
```suggestion
        """Helper function that extracts the predictions from a SpaCy model.
        
        Aside from extracting the predictions from the model, it also aligns the
        predictions with the gold tokens, in case the SpaCy tokeniser tokenises the
        text different from those.
        
        Args:
            tokens_processed (tuple):
                A pair of the labels, being a list of strings, and the SpaCy processed
                document, being a Spacy `Doc` instance.
                
        Returns:
            list:
                A list of predictions for each token, of the same length as the gold
                tokens (first entry of `tokens_processed`).
        """
```